Add new image normalization classes

larrybradley opened the pull request:
This PR contains the following changes to the `visualization` subpackage:

* ~~Adds a `Norm` class to generate a `matplotlib.colors.Normalize` object from an interval and a stretch instance.~~
* ~~Adds a convenience `ImgNorm` class for quick interactive analysis (and used by the `fits2bitmap` command-line script).  `ImgNorm` replaces the `scale_image` function.~~
* Deprecates the `scale_image` function (in favor of the `ImgNorm` class).
* Adds a default stretch for the existing `Normalization` class.
* Adds default `vmin/vmax` for the `ManualInterval` class.
* Adds missing API docs for existing interval and stretch classes.
* Adds addition tests for the visualization subpackage.
* Augments and updates the visualization narrative docs.

Examples
----------

With the existing `ImageNormalize` class one needs to input the `vmin/vmax` values, which is a bit cumbersome, e.g.:

```
interval = MinMaxInterval()
vmin, vmax = interval.get_limits(image)
norm = ImageNormalize(vmin=vmin, vmax=vmax, stretch=SqrtStretch())
```

With the new ~~`Norm` class~~ inputs, one can simply do, e.g.:

```
norm = ImageNormalize(image, MinMaxInterval(), SqrtStretch())
```

And then the image is displayed with:
```
ax.imshow(image, origin='lower', norm=norm)
```

cc: @astrofrog

astrofrog commented:
@larrybradley - I'm in favor of having conveniences such as these, but I think that having ``Normalize``, ``Norm``, and ``ImgNorm`` is confusing - instead I would propose making the two new classes you are adding here class methods for ``Normalize``. In fact, ``Norm`` could even be merged with the existing ``Normalize`` initializer so that you can either specify ``vmin`` and ``vmax`` or ``interval`` and ``image`` in ``__init__``.

``ImgNorm`` can then be made into a convenience class method (I'm trying to think of a good name).

astrofrog commented on astropy/visualization/stretch.py:
```diff
@@ -47,38 +47,36 @@ def _prepare(values, out=None, clip=True):
     """
 
-    def __call__(self, values, out=None, clip=True):
+    def __call__(self, values, clip=True, out=None):
```
Just out of curiosity, why the change?

larrybradley replied:
To make it consistent with the `BaseInterval` `__call__` method.  Also, the underlying `numpy` functions have `out` as the last keyword.

If you agree with `out` being last, then I should also change the calling sequence in the `stretch._prepare` function, so that everything is consistent.

MSeifert04 replied:
should this be mentioned in the changelog?

larrybradley replied:
I suspect that not many people were actually calling stretch instances directly on the data (with or without the `out` keyword), but I'm happy to mention it in the changelog.  @astrofrog?

pllim commented:
> trying to think of a good name

`ConvenientNorm`? `NormForLazyPeople`? `EasyNorm`? `QuickNorm`?

astrofrog commented:
@pllim - just to be clear, I meant a class method not class factory (so ``Normalize.from_...``)

pllim commented:
`from_image`?

larrybradley commented:
@astrofrog  Indeed, I was struggling with short-ish names (because this is a convenience class) similar to `ImageNormalize`.  Suggestions welcome!

larrybradley commented:
@astrofrog  I assume you mean the `ImageNormalize` class (in astropy visualization), not the `Normalize` class (in matplotlib).

I don't think having these new classes as `ImageNormalize` methods is very convenient.  It's also very confusing.  Is something like this what you were thinking:

```
norm = ImageNormalize(stretch=SqrtStretch())
norm2 = norm.apply_interval(data, interval=MinMaxInterval())
imshow(data, norm=norm2)
```

I didn't want to change `ImageNormalize`, but it could be combined with `Norm`.  I think that is *much* better and less confusing than adding methods (and it requires one less unique class name :smile:).  It would require a bit of input validation checking to make sure the right combination of keywords are inputs.  Sound OK?

We'd still need a new name for `ImgNorm`.  This class essentially replaces the `scale_image` function, but `scale` can mean so many things, so it's not the best name.

astrofrog commented:
> I didn't want to change ImageNormalize, but it could be combined with Norm. I think that is much better and less confusing than adding methods (and it requires one less unique class name 😄). It would require a bit of input validation checking to make sure the right combination of keywords are inputs. Sound OK?

Yep!

larrybradley commented:
@astrofrog I've implemented the new changes:

* `ImageNormalize` can be initialized with either `data`/`interval` or `vmin`/`vmax` (which takes precedence).

* Replaced `scale_image` (deprecated) with a new function called `mpl_norm` that returns an `ImageNormalize` object.  (The `ImgNorm` class from the first implementation of this PR is gone).

astrofrog commented on CHANGES.rst:
```diff
@@ -29,9 +29,6 @@ New Features
 
-  - Add a class to combine astropy fitters and functions to remove outliers
-    e. g., sigma clip. [#4760]
-
```
Is there a reason why you are removing this?

larrybradley replied:
Not sure how that happened.  Thanks for catching it.  It's fixed.

astrofrog commented on CHANGES.rst:
```diff
@@ -160,24 +170,17 @@ Bug Fixes
-    deprecated, since this class does not allow stacks of matrices.  As a
-    result, the semi-public functions ``angles.rotation_matrix`` and
-    ``angles.angle_axis`` are also deprecated, in favour of the new routines
-    with the same name in ``coordinates.matrix_utilities``. [#5104]
```
Same here - looks like an issue with conflict resolution?

larrybradley replied:
Fixed.

astrofrog commented:
@larrybradley - thanks for your work on this. I think ``simple_image_norm`` would be better than ``mpl_norm`` - changing ``mpl`` to ``image`` because it returns an ``ImageNormalization`` class, and adding ``simple`` to make it clearer that this is a convenience function that only supports a subset of cases.

larrybradley commented:
@astrofrog Would `image_norm` be acceptable instead?  I understand the reasoning behind prepending `simple`, but that is explained in the docs.

larrybradley commented:
@astrofrog  I renamed the function `simple_norm`.

astrofrog commented:
This all looks good to me now - thanks! Is this good to go from your side?

larrybradley commented:
Yep, it's good to go.

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -6,39 +6,73 @@
 """
 
 from __future__ import division, print_function
-
```
why remove the blank line?

I always thought they were meant to seperate builtins, externals and internal packages

larrybradley replied:
I'm not a fan of the blank lines within the imports.  I wasn't aware that was recommended.

MSeifert04 replied:
I'm not sure if it's really recommended. [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) states they should be grouped and I always interpreted this as blank lines. :smile:

larrybradley replied:
I do order them, just without blank lines. :)  Ah, OK.  `pep8` has failed me all these years (it could at least check the standard library imports).

pllim replied:
Blank line also makes it stands out, so easier to remove when no longer needed.

larrybradley replied:
@pllim But the `__future__` import always goes first, so it's hard to miss! 😄

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -6,39 +6,73 @@
         else:
             if out.dtype.kind != 'f':
-                raise TypeError("Can only do in-place scaling for floating-point arrays")
+                raise TypeError('Can only do in-place scaling for '
```
Somwhere it was mentioned that error messages should start with a lowercase. Don't remember where though.

larrybradley replied:
I always write them as sentences (what if the message contains multiple sentences?).  In any case, @astrofrog actually wrote this one, I just wrapped the line.

MSeifert04 replied:
At least the python-builtin exceptions start lowercase. I think the rational was that the sentence starts with the capitalized exception name and then the first sentence naturally would be lowercase.

I don't mind the capitalized start so feel free to ignore this.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -1,14 +1,18 @@
 """
 
 from __future__ import division, print_function
-
```
same question here regarding the line removal.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -1,14 +1,18 @@
 
 try:
-    import matplotlib  # pylint: disable=W0611
+    import matplotlib    # pylint: disable=W0611
```
why increase the number of whitespaces? two are recommended before inline comments.

larrybradley replied:
I think it's *at least* two spaces.  I think this was inadvertent, so I could change it back.

MSeifert04 replied:
You are right it's at least 2 spaces.

MSeifert04 commented on astropy/visualization/stretch.py:
```diff
@@ -6,20 +6,19 @@
 """
 
 from __future__ import division, print_function
-
```
same question here regarding the blank lines.

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -90,9 +129,9 @@ class AsymmetricPercentileInterval(BaseInterval):
-        from the array (with replacement)
+        Maximum number of values to use. If this is specified, and there
+        are more values in the dataset as this, then values are randomly
+        sampled from the array (with replacement)
```
Missing dot at the end of the sentence.

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -130,15 +168,17 @@ class PercentileInterval(AsymmetricPercentileInterval):
-        from the array (with replacement)
+        Maximum number of values to use. If this is specified, and there
+        are more values in the dataset as this, then values are randomly
+        sampled from the array (with replacement)
```
same here (missing `.`)

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -152,23 +192,23 @@ class ZScaleInterval(BaseInterval):
-        Number of points in array to sample for determining scaling factors.
-        Default to 1000.
+        Number of points in array to sample for determining scaling
+        factors.  Default to 1000.
```
I don't think `Default to ...` is right. Shouldn't it be `DefaultS to ...` or `Default IS ...`

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -152,23 +192,23 @@ class ZScaleInterval(BaseInterval):
-        display. Default to 0.25.
+        Scaling factor (between 0 and 1) for determining min and max.
+        Larger values increase the difference between min and max values
+        used for display. Default to 0.25.
```
same here (default to ...)

MSeifert04 commented on astropy/visualization/interval.py:
```diff
@@ -152,23 +192,23 @@ class ZScaleInterval(BaseInterval):
-        If more than ``max_reject * npixels`` pixels are rejected, then the
-        returned values are the min and max of the data. Default to 0.5.
+        If more than ``max_reject * npixels`` pixels are rejected, then
+        the returned values are the min and max of the data. Default to
```
and here.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -18,10 +22,11 @@ class Normalize(Normalize, object):
         def __init__(self, *args, **kwargs):
-            raise ImportError("matplotlib is required in order to use this class")
+            raise ImportError('matplotlib is required in order to use this '
+                              'class')
```
missing `.` at the end of the sentence.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -30,24 +35,48 @@ class ImageNormalize(Normalize):
 
     Parameters
     ----------
+    data : `~numpy.ndarray`, optional
```
If it's optional what is the default?

MSeifert04 replied:
okay, I think I understand why giving a default value might be a problem here.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
+    data : `~numpy.ndarray`
+        The image array.
+
+    stretch : {{'linear', 'sqrt', 'power', log', 'asinh'}}, optional
```
why two curly braces? I've never seen this way of giving allowed values.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
+        For ``stretch='asinh'``, the value where the asinh curve
+        transitions from linear to logarithmic behavior, expressed as a
+        fraction of the normalized image.  Must be in the range between
+        0 and 1.
```
Missing default value?

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
+        The pixel value of the minimum cut level.  Data values less than
+        ``min_cut`` will set to ``min_cut`` before stretching the image.
+        The default is the image minimum.  ``min_cut`` overrides
+        ``min_percent``.
```
Missing default value?

MSeifert04 replied:
ups, I can't read. I just spotted it. Ignore this comment.

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
+        The pixel value of the maximum cut level.  Data values greater
+        than ``min_cut`` will set to ``min_cut`` before stretching the
+        image.  The default is the image maximum.  ``max_cut`` overrides
+        ``max_percent``.
```
here too

MSeifert04 commented on astropy/visualization/mpl_normalize.py:
```diff
@@ -86,9 +114,108 @@ def __call__(self, values, clip=None):
+    elif stretch == 'asinh':
+        stretch = AsinhStretch(asinh_a)
+    else:
+        raise ValueError('Unknown stretch: {0}'.format(stretch))
```
missing `.` at the end of the sentence (and maybe the capitalization).

MSeifert04 commented on astropy/visualization/scripts/fits2bitmap.py:
```diff
@@ -28,25 +28,24 @@ def fits2bitmap(filename, ext=0, out_fn=None, scale='linear',
+        For ``stretch='asinh'``, the value where the asinh curve
+        transitions from linear to logarithmic behavior, expressed as a
+        fraction of the normalized image.  Must be in the range between
+        0 and 1.
```
missing default value.

MSeifert04 commented on astropy/visualization/scripts/fits2bitmap.py:
```diff
@@ -95,16 +93,16 @@ def fits2bitmap(filename, ext=0, out_fn=None, scale='linear',
     if cmap not in cm.datad:
-        log.critical('{0} is not a valid matplotlib colormap '
-                     'name'.format(cmap))
+        log.critical('{0} is not a valid matplotlib colormap name'
```
missing `.` at the end of the sentence.

MSeifert04 commented on astropy/visualization/scripts/fits2bitmap.py:
```diff
@@ -95,16 +93,16 @@ def fits2bitmap(filename, ext=0, out_fn=None, scale='linear',
 
-    mimg.imsave(out_fn, image_scaled, cmap=cmap)
+    mimg.imsave(out_fn, norm(image), cmap=cmap)
     log.info('Saved file to {0}'.format(out_fn))
```
missing `.` at the end of the info message.

MSeifert04 commented on astropy/visualization/stretch.py:
```diff
@@ -344,8 +379,11 @@ class HistEqStretch(BaseStretch):
+        The data defining the equalization.
+    values : array-like, optional
+        The input image values, which should already be normalized to
+        the [0:1] range.
```
Default value?

MSeifert04 commented on astropy/visualization/stretch.py:
```diff
@@ -362,22 +400,28 @@ def __init__(self, data, values=None):
+        The data defining the equalization.
+    values : array-like, optional
+        The input image values, which should already be normalized to
+        the [0:1] range.
```
does it make sense to include the default value?

larrybradley replied:
I'm not sure (I don't use the histogram-equalized stretches).

MSeifert04 commented on astropy/visualization/ui.py:
```diff
@@ -1,14 +1,18 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
 import numpy as np
-
```
I would suggest including this blank line again.

MSeifert04 commented:
I'm sorry for the flood of stylistic comments. I don't feel strong about them, so feel free to ignore them.

larrybradley commented:
Thanks, @MSeifert04.  I've implemented everything (except explaining the hist-eq `values` default).

@astrofrog  *Now* this should be ready to go.

astrofrog commented:
Thanks @larrybradley and sorry for the delay in merging!
